[GPU] Extend Vector distribute to multiple reductions

pashu123 opened the pull request:
Extends vector distribution in a dispatch with single reduction kernel
to mulitple reduction kernels.

pashu123 committed:
[GPU] Vector distribution on arbitrary dispatches.

Extends vector distribution in a dispatch with single reduction kernel
to mulitple reduction kernels.

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -270,21 +270,70 @@ static CodeGenPipeline getTensorCorePipeline(Type elementType) {
+/// or one of M, N dim isn't present.
+static bool isMatvecLike(linalg::LinalgOp linalgOp) {
+
+  SmallVector<int64_t> bounds = linalgOp.getStaticLoopRanges();
```
```suggestion
static bool isMatvecLike(linalg::LinalgOp linalgOp) {
  SmallVector<int64_t> bounds = linalgOp.getStaticLoopRanges();
```

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -270,21 +270,70 @@ static CodeGenPipeline getTensorCorePipeline(Type elementType) {
+                                              ArrayRef<unsigned> vec2) {
+  SmallVector<unsigned> intersection;
+  for (unsigned element : vec1) {
+    if (std::find(vec2.begin(), vec2.end(), element) != vec2.end()) {
```
Use `llvm::is_contained`.

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -270,21 +270,70 @@ static CodeGenPipeline getTensorCorePipeline(Type elementType) {
 }
 
+/// Returns the intersection of two vectors.
+static SmallVector<unsigned> findIntersection(ArrayRef<unsigned> vec1,
```
I think you can use `llvm::set_intersect` for this: https://github.com/llvm/llvm-project/blob/50428fb5e90b8318148abdf67eefe99bf8fd310b/llvm/include/llvm/ADT/SetOperations.h#L58

pashu123 replied:
I would prefer the first because of the interchange between vector and the set.

kuhar replied:
I think it would be better to construct a temporary SetVector and use the helper. The reason is that this ADT code is tested and deals with edge cases like duplicate values.

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -375,97 +371,168 @@ setReductionVectorDistributionConfig(IREE::GPU::TargetAttr target,
+                                   ArrayRef<unsigned> sharedWgpDims,
+                                   int64_t workgroupSize, int64_t subgroupSize,
+                                   int64_t threadLoads) {
+  auto context = op.getContext();
```
Since the type is not immediately obvious based on the rhs only:

```suggestion
  MLIRContext* context = op.getContext();
```

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -375,97 +371,168 @@ setReductionVectorDistributionConfig(IREE::GPU::TargetAttr target,
-  for (int64_t dim : parallelDims) {
-    workgroupTileSizes[dim] = 1;
+  // Set the shared wgp dims to 1.
+  for (auto i : sharedWgpDims) {
```
Also here: don't use `auto`

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -375,97 +371,168 @@ setReductionVectorDistributionConfig(IREE::GPU::TargetAttr target,
-    return failure();
+    lastDimWgpTileSize =
+        llvm::APIntOps::GreatestCommonDivisor(
+            {64, uint64_t(parallelSize)}, {64, uint64_t(lastDimWgpTileSize)})
```
Prefer c++ casts (`static_cast`). These are more greppable.

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -375,97 +371,168 @@ setReductionVectorDistributionConfig(IREE::GPU::TargetAttr target,
+      while (lastDimWgpTileSize % (threadBasis * threadLoads) != 0) {
+        threadBasis >>= 1;
+      }
+      subgroupBasis = lastDimWgpTileSize / (threadLoads * threadBasis) == 0
+                          ? 1
+                          : lastDimWgpTileSize / (threadLoads * threadBasis);
```
Hoist the calculation into a local variable.

pashu123 replied:
Please explain.

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -375,97 +371,168 @@ setReductionVectorDistributionConfig(IREE::GPU::TargetAttr target,
+    threadBasis >>= 1;
+
+  int64_t subgroupBasis =
+      partialReductionSize / (threadLoads * threadBasis) == 0
+          ? 1
+          : partialReductionSize / (threadLoads * threadBasis);
```
Also here

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -375,97 +371,168 @@ setReductionVectorDistributionConfig(IREE::GPU::TargetAttr target,
+
+  int64_t threadBasis = 64;
+  while (partialReductionSize % (threadBasis * threadLoads) != 0)
+    threadBasis >>= 1;
```
Missing braces around loop bodies

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -375,97 +371,168 @@ setReductionVectorDistributionConfig(IREE::GPU::TargetAttr target,
-  // divisible by the number of vector loads per thread.
-  if (lastDimReductionSize % threadLoads != 0) {
-    DBGS() << "reduction dim isn't a multiple of threadLoads\n";
+  if (reductionSize % threadLoads != 0)
```
Missing braces

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -479,16 +546,247 @@ setReductionVectorDistributionConfig(IREE::GPU::TargetAttr target,
+  linalg::LinalgOp op = computeOps.front();
+  op.getParallelDims(sharedParallelDims);
+
+  for (auto linalgOp : computeOps) {
```
Also here

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -479,16 +546,247 @@ setReductionVectorDistributionConfig(IREE::GPU::TargetAttr target,
+populateConfigInfo(const llvm::SetVector<linalg::LinalgOp> &computeOps,
+                   IREE::GPU::TargetAttr target, int64_t workgroupSize,
+                   int64_t subgroupSize, int64_t threadLoads) {
+  if (computeOps.empty())
```
Also here

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -479,16 +546,247 @@ setReductionVectorDistributionConfig(IREE::GPU::TargetAttr target,
+    sharedParallelDims = findIntersection(sharedParallelDims, currParallelDims);
+  }
+
+  for (auto linalgOp : computeOps) {
```
Also here

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -479,16 +546,247 @@ setReductionVectorDistributionConfig(IREE::GPU::TargetAttr target,
+    bool reductionOp = hasReductionIterator(linalgOp);
+    for (OpOperand *operand : linalgOp.getDpsInputOperands()) {
+      auto operandIdx = linalgOp.getIndexingMapIndex(operand);
+      auto indexingMap = linalgOp.getIndexingMapsArray()[operandIdx];
```
Also here

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -479,16 +546,247 @@ setReductionVectorDistributionConfig(IREE::GPU::TargetAttr target,
+            linalgOp, target, sharedParallelDims, workgroupSize, subgroupSize,
+            threadLoads);
+        if (failed(loweringConfig))
+          return failure();
```
Also here

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -479,16 +546,247 @@ setReductionVectorDistributionConfig(IREE::GPU::TargetAttr target,
+static FailureOr<SetVector<linalg::LinalgOp>>
+checkDispatchForVectorDistribution(mlir::FunctionOpInterface entryPoint) {
+
+  Operation *storeOp = nullptr;
```
```suggestion
static FailureOr<SetVector<linalg::LinalgOp>>
checkDispatchForVectorDistribution(mlir::FunctionOpInterface entryPoint) {
  Operation *storeOp = nullptr;
```

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -479,16 +546,247 @@ setReductionVectorDistributionConfig(IREE::GPU::TargetAttr target,
+
+  SetVector<linalg::LinalgOp> computeOps;
+  bool containsValidReductionOp = true;
+  for (auto op : llvm::reverse(slice)) {
```
Also this

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -479,16 +546,247 @@ setReductionVectorDistributionConfig(IREE::GPU::TargetAttr target,
+  for (auto op : llvm::reverse(slice)) {
+    if (auto linalgOp = dyn_cast<linalg::LinalgOp>(op)) {
+      if (isa<linalg::FillOp>(op))
+        continue;
```
Also here

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -479,16 +546,247 @@ setReductionVectorDistributionConfig(IREE::GPU::TargetAttr target,
+  // Get the reduction dimensions.
+  auto getReductionDims = [](linalg::LinalgOp &linalgOp) -> SetVector<int64_t> {
+    SetVector<int64_t> reductionDims;
+    for (auto iter : llvm::enumerate(linalgOp.getIteratorTypesArray())) {
```
Also here

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -479,16 +546,247 @@ setReductionVectorDistributionConfig(IREE::GPU::TargetAttr target,
+    return reductionDims;
+  };
+
+  for (auto linalgOp : computeOps) {
```
Also here

kuhar commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -479,16 +546,247 @@ setReductionVectorDistributionConfig(IREE::GPU::TargetAttr target,
+  for (auto linalgOp : computeOps) {
+    for (OpOperand *operand : linalgOp.getDpsInputOperands()) {
+      auto operandIdx = linalgOp.getIndexingMapIndex(operand);
+      auto indexingMap = linalgOp.getIndexingMapsArray()[operandIdx];
```
Also here

kuhar reviewed (commented):
This configuration logic is very tricky for me -- is there any way to split this into a few smaller (and easier to review) PRs?

Commit:
check for matmul like op.

We should really check for skinny matmul rather than matvec. Matvec
would also be handled.

pashu123 committed:
Address comments
